crates/bevy_ecs/src/world/mod.rs

```diff
@@ -959,18 +981,34 @@ impl World {
         unsafe { self.get_resource_unchecked_mut() }
     }
 
-    // PERF: optimize this to avoid redundant lookups
     /// Gets a mutable reference to the resource of type `T` if it exists,
     /// otherwise inserts the resource using the result of calling `func`.
     #[inline]
     pub fn get_resource_or_insert_with<R: Resource>(
         &mut self,
         func: impl FnOnce() -> R,
     ) -> Mut<'_, R> {
-        if !self.contains_resource::<R>() {
-            self.insert_resource(func());
+        let change_tick = self.change_tick();
+        let last_change_tick = self.last_change_tick();
+
+        let component_id = self.components.init_resource::<R>();
+        let data = self.initialize_resource_internal(component_id);
+        if !data.is_present() {
+            OwningPtr::make(func(), |ptr| {
```

Comment:
This is interesting and probably isn't the right place to discuss this, but if func panics here. It leaves the World in an invalid state: the ComponentInfo exists for R, but the corresponding ResourceData does not. This isn't an issue in the original implementation.

This problem is typically something I don't think we consider all that much in the ECS and probably worth auditing our use of callbacks.

Reply:
Wait brainfart: the ResourceData exists but it's not populated. That isn't an invalid state.